File: python/src/day8/main.py
from src.common.file_utils import get_path, read_lines


def part_one(filename: str) -> int:
    return run_code(read_lines(get_path(__file__, filename)))[1]


def part_two(filename: str) -> int:
    lines = read_lines(get_path(__file__, filename))

    for i, line in enumerate(lines):
        code, value = line.split(" ")

        if code == "nop":
            line_copy = lines[:]
            line_copy[i] = "{} {}".format("jmp", value)
            completed, acc = run_code(line_copy)
            if completed:
                return acc

        elif code == "jmp":
            line_copy = lines[:]
            line_copy[i] = "{} {}".format("nop", value)
            completed, acc = run_code(line_copy)
            if completed:
                return acc

    return -1


def run_code(lines) -> (bool, int):
    acc, idx = 0, 0
    visited = set()

    while idx < len(lines) and idx not in visited:
        visited.add(idx)
        line = lines[idx]
        code, value = line.split(" ")

        if code == "acc":
            acc = acc + \
                int(value[1:]) if value[0] == "+" else acc - int(value[1:])
            idx += 1

        elif code == "jmp":
            idx = idx + \
                int(value[1:]) if value[0] == "+" else idx - int(value[1:])

        elif code == "nop":
            idx += 1

    return (idx not in visited, acc)


if __name__ == '__main__':

    print("---Part One---")
    part_one_res = part_one("input.txt")
    print(part_one_res)

    print("---Part Two---")
    part_two_res = part_two("input.txt")
    print(part_two_res)

File: go/day8/main.go
package main

import (
	input "aoc2020/inpututils"

	"fmt"
	"strconv"
	"strings"
)

func main() {
	fmt.Println("--- Part One ---")
	fmt.Println(Part1("input.txt"))

	fmt.Println("--- Part Two ---")
	fmt.Println(Part2("input.txt"))
}

func Part1(filename string) int {
	lines := input.ReadLines(filename)
	_, acc := runCode(lines)
	return acc
}

func Part2(filename string) int {
	lines := input.ReadLines(filename)
	
	for i, line := range(lines){
		split := strings.Split(line, " ")
		code, val := split[0], split[1]
		copy := copySlice(lines)

		if code == "nop" {
			newSlice := append(append(copy[:i], fmt.Sprintf("%s %s", "jmp", val)), copy[i+1:]...)
			completed, acc := runCode(newSlice)
			if completed {
				return acc
			}
		}
		
		if code == "jmp" {
			newSlice := append(append(copy[:i], fmt.Sprintf("%s %s", "nop", val)), copy[i+1:]...)
			completed, acc := runCode(newSlice)
			if completed {
				return acc
			}
		}

	}
	return -1
}

func runCode(lines []string) (bool, int) {
	acc, idx := 0, 0
	visited := map[int]bool{}

	for idx < len(lines) && !visited[idx] {
		visited[idx] = true

		split := strings.Split(lines[idx], " ")
		code, val := split[0], split[1]

		if code == "acc" {
			acc += getValueFromString(val)
			idx += 1
		} else if code == "jmp" {
			idx += getValueFromString(val)
		} else if code == "nop" {
			idx += 1
		}
	
	}
	return idx >= len(lines), acc
}

func copySlice(a []string) []string{
	return append(a[:0:0], a...)
}

func getValueFromString(s string) int {
	if string(s[0]) == "-" {
		return -1 * toInt(s[1:])
	}
	return toInt(s[1:])
}

func check(err error) {
	if err != nil {
		panic(err)
	}
}

func toInt(s string) int {
	i, err := strconv.Atoi(s)
	check(err)
	return i
}
